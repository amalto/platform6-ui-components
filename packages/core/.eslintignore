dist/
loader/

src/components.d.ts