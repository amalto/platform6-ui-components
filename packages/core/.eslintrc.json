{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
