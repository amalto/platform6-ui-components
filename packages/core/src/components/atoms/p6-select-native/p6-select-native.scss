@import 'src/styles/atoms/control';

:host {
  display: inline-block;

  .select {
    display: inline-block;
    max-width: 100%;
    position: relative;
    vertical-align: top;

    &:not(.is-multiple) {
      height: $input-height;
    }

    &:not(.is-multiple):not(.is-loading) {
      &::after {
        @include arrow($input-arrow);

        border-width: 2px;
        right: 1.125em;
        z-index: 4;
      }
    }

    &.is-fullwidth {
      width: 100%;
    }

    select {
      &:not([multiple]) {
        padding-right: 2.5em;
      }

      &[disabled]:hover {
        border-color: $input-disabled-border-color;
      }

      @include input;
      @include select;

      cursor: pointer;
      display: block;
      font-size: 1em;
      max-width: 100%;
      outline: none;

      &::-ms-expand {
        display: none;
      }

      &[multiple] {
        height: auto;
        padding: 0;

        option {
          padding: 0.5em 1em;
        }
      }
    }

    &.is-fullwidth select {
      width: 100%;
    }

    select fieldset[disabled] &:hover {
      border-color: $input-disabled-border-color;
    }

    // States
    &:not(.is-multiple):not(.is-loading):hover {
      &::after {
        border-color: $input-hover-color;
      }
    }

    // Colors
    @each $name, $pair in $colors {
      $color: nth($pair, 1);
      &.is-#{$name} {
        &:not(:hover)::after {
          border-color: $color;
        }

        select {
          border-color: $color;

          &:hover,
          &.is-hovered {
            border-color: bulmaDarken($color, 5%);
          }

          &:focus,
          &.is-focused,
          &:active,
          &.is-active {
            box-shadow: $input-focus-box-shadow-size bulmaRgba($color, 0.25);
          }
        }
      }
    }
    // Sizes
    @include control-normal;

    // Modifiers
    &.is-disabled {
      &::after {
        border-color: $input-disabled-color;
      }
    }

    &.is-loading {
      &::after {
        @include loader;

        margin-top: 0;
        position: absolute;
        right: 0.625em;
        top: 0.625em;
        transform: none;
      }

      &.is-small::after {
        font-size: $size-small;
      }
    }
  }
}
