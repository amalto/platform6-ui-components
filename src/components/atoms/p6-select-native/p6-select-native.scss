@import "bulma/sass/form/select";

:host {
  display: block;
}
