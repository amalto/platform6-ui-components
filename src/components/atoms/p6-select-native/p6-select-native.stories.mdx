import {
  Story,
  Preview,
  Meta,
  Markdown,
  Description,
} from "@storybook/addon-docs/blocks";

import Readme from "./readme.md";

import {
  DefaultStory,
  SelectedStory,
  DisabledStory,
  SizeStory,
  ModeStory,
  Previews,
} from "./p6-select-native.stories.ts";

<Meta
  title="Atoms/Select native"
  component="p6-select-native"
  parameters={{
    actions: {
      handles: ["p6Submit p6-form"],
    },
  }}
/>

# Select

`p6-select-native` provides a select

<Preview>
  <Story name="Default" height="58px" parameters={Previews.Default()}>
    {DefaultStory()}
  </Story>
</Preview>

## Selected value

<Preview>
  <Story name="Selected value" height="58px" parameters={Previews.Selected}>
    {SelectedStory()}
  </Story>
</Preview>

## Disabled

<Preview>
  <Story
    name="Disabled"
    height="58px"
    parameters={Previews.Default({ disabled: true })}
  >
    {DisabledStory()}
  </Story>
</Preview>

## Sizes

<Preview>
  <Story
    name="Sizes"
    height="58px"
    parameters={Previews.Default({ size: "small" })}
  >
    {SizeStory()}
  </Story>
</Preview>

## Modes

<Preview>
  <Story
    name="Modes"
    height="58px"
    parameters={Previews.Default({ mode: "success" })}
  >
    {ModeStory()}
  </Story>
</Preview>

<Description markdown={Readme} />
