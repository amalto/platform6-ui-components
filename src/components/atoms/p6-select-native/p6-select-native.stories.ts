import { Components } from '../../../components';
import { ComponentProps, getElement, getForm, makeModeStory, makeStory, Props } from '../../../shared/storybook';
import { Mode } from '../../../shared/types';

const component = 'p6-select-native';

export default {
  title: 'Atoms/Select native',
  component,
};

const componentProps: ComponentProps = ['name', 'multiple', 'mode', 'fullWidth', 'disabled', 'required', 'readOnly'];

const getStoryField = (options: HTMLElement[], props?: Props<Components.P6SelectNative>): HTMLElement =>
  getElement(component, options, {
    name: 'language',
    ...props,
  });

const getOption = (value: string, display: string, props?: Props<HTMLOptionElement>): HTMLElement => {
  return getElement('option', display, {
    value,
    ...props,
  });
};

export const Default = makeStory<{
  disabled: boolean;
  readOnly: boolean;
  required: boolean;
  fullWidth: boolean;
  multiple: boolean;
  mode: Mode;
}>({
  componentProps,
  args: {
    disabled: false,
    readOnly: false,
    required: false,
    fullWidth: false,
    multiple: false,
    mode: Mode.default,
  },
  builder: (args): HTMLElement =>
    getStoryField([getOption('', 'Select language'), getElement('optgroup', [getOption('fr', 'Français'), getOption('en', 'English')], { label: 'Europe' })], args),
});

export const Selected = makeStory({
  componentProps,
  builder: (): HTMLElement => getStoryField([getOption('', 'Placeholder'), getOption('value', 'Selected value', { selected: true })]),
});

export const Disabled = makeStory<{ disabled: boolean }>({
  componentProps,
  args: { disabled: true },
  builder: (props): HTMLElement => getStoryField([getOption('', 'Placeholder'), getOption('value', 'Display')], props),
});

export const Modes = makeModeStory({
  componentProps,
  builder: ({ key, value }) =>
    getStoryField([getOption('', 'Placeholder'), getOption('value', key)], {
      mode: value,
    }),
});

export const Form = makeStory<{
  disabled: boolean;
  readOnly: boolean;
  required: boolean;
  multiple: boolean;
}>({
  componentProps,
  args: {
    disabled: false,
    readOnly: false,
    required: false,
    multiple: false,
  },
  builder: (props): HTMLElement =>
    getForm(
      getStoryField(
        [getOption('', 'Select language'), getElement('optgroup', [getOption('fr', 'Français'), getOption('en', 'English', { selected: true })], { label: 'Europe' })],
        { name: 'field', ...props },
      ),
    ),
});
