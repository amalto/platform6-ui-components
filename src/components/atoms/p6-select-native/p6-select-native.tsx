import {
  Component,
  ComponentInterface,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Method,
  Prop,
} from "@stencil/core";
import { P6Control } from "~shared/form/control";
import { InvalidEventDetail, ValidEventDetail } from "~shared/form/event";
import {
  defaultCheckValidity,
  defaultValidationMessage,
} from "~shared/form/validation";
import { Mode, Size } from "~shared/types";
import { getModeClass, getSizeClass } from "~utils/classes";
import {
  isHTMLOptGroupElement,
  isHTMLOptionElement,
  toArray,
} from "~utils/dom";

export type P6SelectValue = string[] | string | undefined;
export type P6SelectControl = P6Control<P6SelectValue>;

@Component({
  tag: "p6-select-native",
  styleUrl: "p6-select-native.scss",
  shadow: true,
})
export class P6SelectNative implements ComponentInterface, P6SelectControl {
  /**
   * The name
   */
  @Prop() name!: string;

  /**
   * Is multiple
   */
  @Prop() multiple = false;

  /**
   * The size of the component to display
   */
  @Prop() size: Size = Size.normal;

  /**
   * The Mode of the component to display
   */
  @Prop() mode: Mode = Mode.default;

  /**
   * The select should take the full width
   */
  @Prop({ attribute: "fullWidth" }) fullWidth = false;

  /**
   * The select is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * Marks the select as required. It can't be submitted without a value
   */
  @Prop() required = false;

  /**
   * Marks as read only.
   */
  @Prop({ attribute: "readOnly" }) readOnly = false;

  /**
   * When the select is valid
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6SelectValue>>;

  /**
   * When the select is invalid
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  /**
   * When the select ask to register
   */
  @Event() p6FormRegister!: EventEmitter<P6SelectControl>;

  /**
   * When the select ask to unregister
   */
  @Event() p6FormUnregister!: EventEmitter<P6SelectControl>;

  /**
   * When the value change
   */
  @Event() p6Change!: EventEmitter<ValidEventDetail<P6SelectValue>>;

  @Element() host!: HTMLP6SelectNativeElement;

  private nativeInput: HTMLSelectElement | undefined;

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    await defaultCheckValidity<P6SelectValue>({
      name: this.name,
      disabled: this.disabled,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      getValue: () => this.inputValue,
    });

    return Promise.resolve(true);
  }

  render(): JSX.Element {
    return (
      <Host
        aria-disabled={this.disabled ? "true" : null}
        required={this.required}
      >
        <div
          class={{
            select: true,
            "is-fullwidth": this.fullWidth,
            ...getSizeClass(this.size),
            ...getModeClass(this.mode),
          }}
        >
          <select
            name={this.name}
            disabled={this.disabled}
            multiple={this.multiple}
            onBlur={this.onChangeHandler}
            ref={(input): void => {
              this.nativeInput = input;
            }}
          />
        </div>
      </Host>
    );
  }

  private get inputValue(): string | string[] | undefined {
    if (this.nativeInput === undefined) {
      return undefined;
    }

    const selectedOptions =
      undefined === this.nativeInput
        ? []
        : toArray<HTMLOptionElement, HTMLCollectionOf<HTMLOptionElement>>(
            this.nativeInput.selectedOptions
          );
    const value = selectedOptions.map((option) => option.value);

    return this.multiple ? value : value.shift();
  }

  private onChangeHandler = (): void => {
    this.p6Change.emit({
      name: this.name,
      value: this.inputValue,
    });
  };

  componentWillLoad(): Promise<void> | void {
    this.computeOptions();
  }

  private getOptions(element: Element): HTMLOptionElement[] {
    const subItems = toArray(element.children);
    const options = subItems.filter(isHTMLOptionElement);
    subItems
      .filter(isHTMLOptGroupElement)
      .forEach((subItem) => options.push(...this.getOptions(subItem)));
    return options;
  }

  private computeOptions(): void {
    const options = this.getOptions(this.host);

    if (options.length === 0) {
      return;
    }

    if (options.find((option) => option.selected) === undefined) {
      const emptyOption = options.find((child) => child.value === "");
      if (emptyOption !== undefined) {
        emptyOption.selected = true;
      } else {
        options[0].selected = true;
      }
    }

    if (this.readOnly) {
      options
        .filter((option) => !option.selected)
        .forEach((child) => {
          child.setAttribute("disabled", "true");
        });
    }
  }

  componentDidRender(): void {
    toArray(this.host.children).forEach((child) =>
      this.nativeInput?.appendChild(child)
    );
  }

  componentDidLoad(): void {
    this.p6FormRegister.emit(this);
  }

  disconnectedCallback(): void {
    this.p6FormUnregister.emit(this);
  }
}
