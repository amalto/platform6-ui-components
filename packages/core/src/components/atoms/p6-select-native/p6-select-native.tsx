import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, ValidEventDetail } from '../../../shared/form/event';
import { defaultCheckValidity, defaultValidationMessage } from '../../../shared/form/validation';
import { Mode } from '../../../shared/types';
import { isHTMLOptGroupElement, isHTMLOptionElement, toArray } from '../../../utils';
import { getModeClass } from '../../../utils/classes';

export type P6SelectValue = string[] | string | undefined;
export type P6SelectControl = P6Control<P6SelectValue>;

@Component({
  tag: 'p6-select-native',
  styleUrl: 'p6-select-native.scss',
  shadow: true,
})
export class P6SelectNative implements ComponentInterface, P6SelectControl {
  /**
   * The name
   */
  @Prop() name!: string;

  /**
   * Is multiple
   */
  @Prop() multiple = false;

  /**
   * The Mode of the component to display
   */
  @Prop() mode: Mode = Mode.default;

  /**
   * The select should take the full width
   */
  @Prop({ attribute: 'fullWidth' }) fullWidth = false;

  /**
   * The select is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * Marks the select as required. It can't be submitted without a value
   */
  @Prop() required = false;

  /**
   * Marks as read only.
   */
  @Prop({ attribute: 'readOnly' }) readOnly = false;

  /**
   * When the select is valid
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6SelectValue>>;

  /**
   * When the select is invalid
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  /**
   * When the select ask to register
   */
  @Event() p6FormRegister!: EventEmitter<P6SelectControl>;

  /**
   * When the select ask to unregister
   */
  @Event() p6FormUnregister!: EventEmitter<P6SelectControl>;

  /**
   * When the value change
   */
  @Event() p6Change!: EventEmitter<ValidEventDetail<P6SelectValue>>;

  @Element() host!: HTMLP6SelectNativeElement;

  private nativeInput: HTMLSelectElement | undefined;

  private defaultValue: string[] | undefined;

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    await defaultCheckValidity<P6SelectValue>({
      name: this.name,
      disabled: this.disabled,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      getValue: () => this.inputValue,
    });

    return Promise.resolve(true);
  }

  /**
   * Restores the input's default value
   */
  @Method()
  async reset(): Promise<boolean> {
    const result = false;

    if (this.nativeInput !== undefined) {
      this.getOptions(this.nativeInput).forEach(option => {
        // eslint-disable-next-line no-param-reassign
        option.selected = !!this.defaultValue?.includes(option.value);
      });
    }

    return Promise.resolve(result);
  }

  render(): JSX.Element {
    return (
      <Host aria-disabled={this.disabled ? 'true' : null} required={this.required}>
        <div
          class={{
            'select': true,
            'is-fullwidth': this.fullWidth,
            ...getModeClass(this.mode),
          }}
        >
          <select
            name={this.name}
            disabled={this.disabled}
            multiple={this.multiple}
            onBlur={this.onChangeHandler}
            ref={(input): void => {
              this.nativeInput = input;
            }}
          />
        </div>
      </Host>
    );
  }

  private get inputValue(): string | string[] | undefined {
    if (this.nativeInput?.selectedOptions === undefined) {
      return undefined;
    }

    const value = toArray<HTMLOptionElement>(this.nativeInput.selectedOptions).map(option => option.value);

    return this.multiple ? value : value.shift();
  }

  private onChangeHandler = (): void => {
    this.p6Change.emit({
      name: this.name,
      value: this.inputValue,
    });
  };

  componentWillLoad(): Promise<void> | void {
    this.computeOptions();
  }

  private getOptions(element: Element): HTMLOptionElement[] {
    const subItems = toArray(element.children);
    const options = subItems.filter(isHTMLOptionElement);
    subItems.filter(isHTMLOptGroupElement).forEach(subItem => options.push(...this.getOptions(subItem)));
    return options;
  }

  private computeOptions(): void {
    const options = this.getOptions(this.host);

    if (options.length === 0) {
      return;
    }

    if (options.find(option => option.selected) === undefined) {
      const emptyOption = options.find(child => child.value === '');
      if (emptyOption !== undefined) {
        emptyOption.selected = true;
      } else {
        options[0].selected = true;
      }
    }

    if (this.readOnly) {
      options
        .filter(option => !option.selected)
        .forEach(child => {
          child.setAttribute('disabled', 'true');
        });
    }
  }

  componentDidRender(): void {
    toArray(this.host.children).forEach(child => this.nativeInput?.appendChild(child));
  }

  componentDidLoad(): void {
    if (this.defaultValue === undefined) {
      const selectedValues = this.inputValue;
      // eslint-disable-next-line no-nested-ternary
      this.defaultValue = Array.isArray(selectedValues) ? selectedValues : selectedValues === undefined ? [] : [selectedValues];
    }

    this.p6FormRegister.emit(this);
  }

  disconnectedCallback(): void {
    this.p6FormUnregister.emit(this);
  }
}
