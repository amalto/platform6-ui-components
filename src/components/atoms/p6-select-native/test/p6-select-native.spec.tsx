import { newSpecPage } from '@stencil/core/testing';
import { P6SelectNative } from '../p6-select-native';

describe('p6-select-native', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6SelectNative],
      html: `<p6-select-native>
      </p6-select-native>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
