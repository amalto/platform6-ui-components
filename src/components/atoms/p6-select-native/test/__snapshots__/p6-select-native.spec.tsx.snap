// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-select-native renders 1`] = `
<p6-select-native>
  <mock:shadow-root>
    <div class="select">
      <select></select>
    </div>
  </mock:shadow-root>
</p6-select-native>
`;
